File: data_service/backtesting/performance.py
"""
Performance metrics for backtesting and live monitoring.

Implements the standard quant finance KPI suite:
Sharpe, Sortino, Calmar, max drawdown, alpha/beta, etc.
"""

import logging
from typing import Dict, Optional

import numpy as np
import pandas as pd

logger = logging.getLogger(__name__)

TRADING_DAYS = 252


class PerformanceMetrics:
    """
    Compute a comprehensive set of risk/return metrics from a returns series.

    Parameters
    ----------
    returns : pd.Series         Daily portfolio returns.
    benchmark : pd.Series       Daily benchmark returns (optional).
    risk_free_rate : float      Annualised risk-free rate.
    """

    def __init__(
        self,
        returns: pd.Series,
        benchmark: Optional[pd.Series] = None,
        risk_free_rate: float = 0.04,
    ):
        self.returns = returns.dropna()
        self.benchmark = benchmark.reindex(self.returns.index).fillna(0) if benchmark is not None else None
        self.risk_free_rate = risk_free_rate
        self._rf_daily = risk_free_rate / TRADING_DAYS

    def compute(self) -> Dict[str, float]:
        """Return the full metrics dictionary."""
        r = self.returns
        metrics: Dict[str, float] = {}

        # Return
        metrics["total_return"] = self._total_return(r)
        metrics["annualised_return"] = self._annualised_return(r)
        metrics["cagr"] = self._cagr(r)

        # Risk
        metrics["annualised_vol"] = self._annualised_vol(r)
        metrics["downside_vol"] = self._downside_vol(r)
        metrics["max_drawdown"] = self._max_drawdown(r)
        metrics["avg_drawdown"] = self._avg_drawdown(r)
        metrics["var_95"] = self._var(r, 0.05)
        metrics["cvar_95"] = self._cvar(r, 0.05)
        metrics["skewness"] = float(r.skew())
        metrics["kurtosis"] = float(r.kurtosis())

        # Risk-adjusted
        metrics["sharpe"] = self._sharpe(r)
        metrics["sortino"] = self._sortino(r)
        metrics["calmar"] = self._calmar(r)
        metrics["omega"] = self._omega(r)

        # Win/loss
        metrics["win_rate"] = float((r > 0).mean())
        metrics["profit_factor"] = self._profit_factor(r)
        metrics["avg_win"] = float(r[r > 0].mean()) if (r > 0).any() else 0.0
        metrics["avg_loss"] = float(r[r < 0].mean()) if (r < 0).any() else 0.0

        # Benchmark-relative (if provided)
        if self.benchmark is not None:
            metrics["alpha"] = self._alpha(r, self.benchmark)
            metrics["beta"] = self._beta(r, self.benchmark)
            metrics["information_ratio"] = self._information_ratio(r, self.benchmark)
            metrics["tracking_error"] = self._tracking_error(r, self.benchmark)
            metrics["up_capture"] = self._capture(r, self.benchmark, up=True)
            metrics["down_capture"] = self._capture(r, self.benchmark, up=False)

        # Streak
        metrics["max_consec_wins"] = self._max_consecutive(r, positive=True)
        metrics["max_consec_losses"] = self._max_consecutive(r, positive=False)

        return {k: round(float(v), 6) for k, v in metrics.items()}

    # ------------------------------------------------------------------
    # Return metrics
    # ------------------------------------------------------------------

    @staticmethod
    def _total_return(r: pd.Series) -> float:
        return float((1 + r).prod() - 1)

    @staticmethod
    def _annualised_return(r: pd.Series) -> float:
        n = len(r)
        return float((1 + r).prod() ** (TRADING_DAYS / max(n, 1)) - 1)

    @staticmethod
    def _cagr(r: pd.Series) -> float:
        n_years = len(r) / TRADING_DAYS
        total = (1 + r).prod()
        return float(total ** (1 / max(n_years, 1e-6)) - 1)

    # ------------------------------------------------------------------
    # Risk metrics
    # ------------------------------------------------------------------

    @staticmethod
    def _annualised_vol(r: pd.Series) -> float:
        return float(r.std() * np.sqrt(TRADING_DAYS))

    def _downside_vol(self, r: pd.Series) -> float:
        downside = r[r < self._rf_daily] - self._rf_daily
        return float(downside.std() * np.sqrt(TRADING_DAYS)) if len(downside) > 1 else 0.0

    @staticmethod
    def _max_drawdown(r: pd.Series) -> float:
        cum = (1 + r).cumprod()
        peak = cum.cummax()
        dd = (cum - peak) / peak
        return float(dd.min())

    @staticmethod
    def _avg_drawdown(r: pd.Series) -> float:
        cum = (1 + r).cumprod()
        peak = cum.cummax()
        dd = (cum - peak) / peak
        return float(dd[dd < 0].mean()) if (dd < 0).any() else 0.0

    @staticmethod
    def _var(r: pd.Series, alpha: float) -> float:
        return float(np.percentile(r, alpha * 100))

    @staticmethod
    def _cvar(r: pd.Series, alpha: float) -> float:
        cutoff = np.percentile(r, alpha * 100)
        return float(r[r <= cutoff].mean())

    # ------------------------------------------------------------------
    # Risk-adjusted metrics
    # ------------------------------------------------------------------

    def _sharpe(self, r: pd.Series) -> float:
        excess = r - self._rf_daily
        std = excess.std()
        return float(excess.mean() / std * np.sqrt(TRADING_DAYS)) if std > 0 else 0.0

    def _sortino(self, r: pd.Series) -> float:
        excess = r - self._rf_daily
        dv = self._downside_vol(r)
        ann_excess = excess.mean() * TRADING_DAYS
        return float(ann_excess / dv) if dv > 0 else 0.0

    def _calmar(self, r: pd.Series) -> float:
        cagr = self._cagr(r)
        mdd = abs(self._max_drawdown(r))
        return float(cagr / mdd) if mdd > 0 else 0.0

    @staticmethod
    def _omega(r: pd.Series, threshold: float = 0.0) -> float:
        gains = r[r > threshold].sum()
        losses = -r[r <= threshold].sum()
        return float(gains / losses) if losses > 0 else np.inf

    # ------------------------------------------------------------------
    # Win/loss metrics
    # ------------------------------------------------------------------

    @staticmethod
    def _profit_factor(r: pd.Series) -> float:
        gross_win = r[r > 0].sum()
        gross_loss = -r[r < 0].sum()
        return float(gross_win / gross_loss) if gross_loss > 0 else np.inf

    @staticmethod
    def _max_consecutive(r: pd.Series, positive: bool) -> float:
        mask = (r > 0) if positive else (r < 0)
        max_streak, current = 0, 0
        for val in mask:
            current = current + 1 if val else 0
            max_streak = max(max_streak, current)
        return float(max_streak)

    # ------------------------------------------------------------------
    # Benchmark-relative
    # ------------------------------------------------------------------

    @staticmethod
    def _beta(r: pd.Series, b: pd.Series) -> float:
        cov = np.cov(r, b)
        return float(cov[0, 1] / cov[1, 1]) if cov[1, 1] > 0 else 0.0

    def _alpha(self, r: pd.Series, b: pd.Series) -> float:
        beta = self._beta(r, b)
        alpha_daily = r.mean() - self._rf_daily - beta * (b.mean() - self._rf_daily)
        return float(alpha_daily * TRADING_DAYS)

    @staticmethod
    def _information_ratio(r: pd.Series, b: pd.Series) -> float:
        active = r - b
        return float(active.mean() / active.std() * np.sqrt(TRADING_DAYS)) if active.std() > 0 else 0.0

    @staticmethod
    def _tracking_error(r: pd.Series, b: pd.Series) -> float:
        return float((r - b).std() * np.sqrt(TRADING_DAYS))

    @staticmethod
    def _capture(r: pd.Series, b: pd.Series, up: bool) -> float:
        mask = b > 0 if up else b < 0
        if not mask.any():
            return 0.0
        return float(r[mask].mean() / b[mask].mean()) if b[mask].mean() != 0 else 0.0

File: backend/src/performance.cpp
#include "performance.hpp"
#include <spdlog/spdlog.h>
#include <nlohmann/json.hpp>

#include <algorithm>
#include <cmath>
#include <numeric>
#include <stdexcept>

namespace estateiq {

PerformanceCalculator::PerformanceCalculator(
    double risk_free_rate,
    const std::vector<double>* benchmark_returns
)
    : risk_free_rate_(risk_free_rate)
    , rf_daily_(risk_free_rate / TRADING_DAYS)
    , benchmark_(benchmark_returns)
{}

// ─── Compute all metrics ──────────────────────────────────────────────────────
Metrics PerformanceCalculator::compute(const std::vector<double>& r) const {
    Metrics m;
    if (r.empty()) return m;

    m.n_bars              = r.size();
    m.total_return        = total_return(r);
    m.annualised_return   = annualised_return(r);
    m.cagr                = cagr(r);
    m.annualised_vol      = annualised_vol(r);
    m.max_drawdown        = max_drawdown(r);
    m.sharpe              = sharpe(r, rf_daily_);
    m.sortino             = sortino(r, rf_daily_);
    m.calmar              = calmar(r);
    m.win_rate            = win_rate(r);
    m.profit_factor       = profit_factor(r);
    m.var_95              = var(r, 0.05);
    m.cvar_95             = cvar(r, 0.05);
    m.skewness            = skewness(r);
    m.kurtosis            = kurtosis(r);

    double ag = 0.0;
    for (auto x : r) if (x > 0) ag += x;
    double al = 0.0;
    for (auto x : r) if (x < 0) al += x;
    long n_pos = std::count_if(r.begin(), r.end(), [](double x){ return x > 0; });
    long n_neg = std::count_if(r.begin(), r.end(), [](double x){ return x < 0; });
    m.avg_win  = n_pos > 0 ? ag / n_pos : 0.0;
    m.avg_loss = n_neg > 0 ? al / n_neg : 0.0;

    if (benchmark_ && !benchmark_->empty()) {
        m.beta              = beta(r, *benchmark_);
        m.alpha             = alpha_ann(r, *benchmark_, rf_daily_);
        m.information_ratio = information_ratio(r, *benchmark_);
        m.tracking_error    = tracking_error(r, *benchmark_);
    }
    return m;
}

Metrics PerformanceCalculator::compute_from_nav(
    const std::vector<std::pair<TimePoint, double>>& nav_history
) const {
    std::vector<double> r;
    r.reserve(nav_history.size());
    for (std::size_t i = 1; i < nav_history.size(); ++i) {
        double prev = nav_history[i-1].second;
        double curr = nav_history[i].second;
        r.push_back(prev > 0 ? (curr / prev - 1.0) : 0.0);
    }
    return compute(r);
}

// ─── Static helpers ───────────────────────────────────────────────────────────
double PerformanceCalculator::total_return(const std::vector<double>& r) {
    double cum = 1.0;
    for (auto x : r) cum *= (1.0 + x);
    return cum - 1.0;
}

double PerformanceCalculator::annualised_return(const std::vector<double>& r) {
    double tr = total_return(r);
    double n  = static_cast<double>(r.size());
    return std::pow(1.0 + tr, TRADING_DAYS / n) - 1.0;
}

double PerformanceCalculator::cagr(const std::vector<double>& r) {
    if (r.empty()) return 0.0;
    double n_years = static_cast<double>(r.size()) / TRADING_DAYS;
    double tr      = total_return(r);
    return std::pow(1.0 + tr, 1.0 / std::max(n_years, 1e-6)) - 1.0;
}

double PerformanceCalculator::annualised_vol(const std::vector<double>& r) {
    if (r.size() < 2) return 0.0;
    double mean = std::accumulate(r.begin(), r.end(), 0.0) / r.size();
    double var  = 0.0;
    for (auto x : r) { double d = x - mean; var += d * d; }
    return std::sqrt(var / (r.size() - 1)) * std::sqrt(TRADING_DAYS);
}

double PerformanceCalculator::max_drawdown(const std::vector<double>& r) {
    double cum = 1.0, peak = 1.0, mdd = 0.0;
    for (auto x : r) {
        cum  *= (1.0 + x);
        peak  = std::max(peak, cum);
        mdd   = std::min(mdd, (cum - peak) / peak);
    }
    return mdd;
}

double PerformanceCalculator::sharpe(const std::vector<double>& r, double rf_daily) {
    if (r.size() < 2) return 0.0;
    std::vector<double> excess(r.size());
    for (std::size_t i = 0; i < r.size(); ++i) excess[i] = r[i] - rf_daily;
    double mean = std::accumulate(excess.begin(), excess.end(), 0.0) / excess.size();
    double var  = 0.0;
    for (auto x : excess) { double d = x - mean; var += d * d; }
    double std_dev = std::sqrt(var / (excess.size() - 1));
    return std_dev > 0 ? mean / std_dev * std::sqrt(TRADING_DAYS) : 0.0;
}

double PerformanceCalculator::sortino(const std::vector<double>& r, double rf_daily) {
    if (r.size() < 2) return 0.0;
    double mean_excess = 0.0;
    double down_var    = 0.0;
    int    n_down      = 0;
    for (auto x : r) {
        mean_excess += x - rf_daily;
        if (x < rf_daily) { double d = x - rf_daily; down_var += d * d; ++n_down; }
    }
    mean_excess /= r.size();
    if (n_down < 2) return 0.0;
    double down_std = std::sqrt(down_var / n_down) * std::sqrt(TRADING_DAYS);
    return down_std > 0 ? mean_excess * TRADING_DAYS / down_std : 0.0;
}

double PerformanceCalculator::calmar(const std::vector<double>& r) {
    double mdd = std::abs(max_drawdown(r));
    return mdd > 0 ? cagr(r) / mdd : 0.0;
}

double PerformanceCalculator::win_rate(const std::vector<double>& r) {
    if (r.empty()) return 0.0;
    long wins = std::count_if(r.begin(), r.end(), [](double x){ return x > 0; });
    return static_cast<double>(wins) / r.size();
}

double PerformanceCalculator::profit_factor(const std::vector<double>& r) {
    double gross_win = 0.0, gross_loss = 0.0;
    for (auto x : r) { if (x > 0) gross_win += x; else gross_loss -= x; }
    return gross_loss > 0 ? gross_win / gross_loss : 1e9;
}

double PerformanceCalculator::var(const std::vector<double>& r, double alpha) {
    if (r.empty()) return 0.0;
    std::vector<double> sorted = r;
    std::sort(sorted.begin(), sorted.end());
    std::size_t idx = static_cast<std::size_t>(alpha * sorted.size());
    return sorted[std::min(idx, sorted.size() - 1)];
}

double PerformanceCalculator::cvar(const std::vector<double>& r, double alpha) {
    if (r.empty()) return 0.0;
    double cutoff = var(r, alpha);
    double sum = 0.0; int cnt = 0;
    for (auto x : r) if (x <= cutoff) { sum += x; ++cnt; }
    return cnt > 0 ? sum / cnt : cutoff;
}

double PerformanceCalculator::skewness(const std::vector<double>& r) {
    if (r.size() < 3) return 0.0;
    double n    = r.size();
    double mean = std::accumulate(r.begin(), r.end(), 0.0) / n;
    double m2   = 0.0, m3 = 0.0;
    for (auto x : r) {
        double d = x - mean;
        m2 += d * d;
        m3 += d * d * d;
    }
    m2 /= n; m3 /= n;
    double sigma = std::sqrt(m2);
    return sigma > 0 ? m3 / (sigma * sigma * sigma) : 0.0;
}

double PerformanceCalculator::kurtosis(const std::vector<double>& r) {
    if (r.size() < 4) return 0.0;
    double n    = r.size();
    double mean = std::accumulate(r.begin(), r.end(), 0.0) / n;
    double m2   = 0.0, m4 = 0.0;
    for (auto x : r) {
        double d = x - mean;
        m2 += d * d;
        m4 += d * d * d * d;
    }
    m2 /= n; m4 /= n;
    return m2 > 0 ? m4 / (m2 * m2) - 3.0 : 0.0;  // excess kurtosis
}

double PerformanceCalculator::beta(const std::vector<double>& r, const std::vector<double>& b) {
    std::size_t n = std::min(r.size(), b.size());
    if (n < 2) return 1.0;
    double mr = 0.0, mb = 0.0;
    for (std::size_t i = 0; i < n; ++i) { mr += r[i]; mb += b[i]; }
    mr /= n; mb /= n;
    double cov = 0.0, var_b = 0.0;
    for (std::size_t i = 0; i < n; ++i) {
        cov   += (r[i] - mr) * (b[i] - mb);
        var_b += (b[i] - mb) * (b[i] - mb);
    }
    return var_b > 0 ? cov / var_b : 0.0;
}

double PerformanceCalculator::alpha_ann(const std::vector<double>& r,
                                        const std::vector<double>& b,
                                        double rf_daily) {
    double b_ = beta(r, b);
    std::size_t n = std::min(r.size(), b.size());
    double mr = 0.0, mb = 0.0;
    for (std::size_t i = 0; i < n; ++i) { mr += r[i]; mb += b[i]; }
    mr /= n; mb /= n;
    double alpha_daily = mr - rf_daily - b_ * (mb - rf_daily);
    return alpha_daily * TRADING_DAYS;
}

double PerformanceCalculator::information_ratio(const std::vector<double>& r,
                                                 const std::vector<double>& b) {
    std::size_t n = std::min(r.size(), b.size());
    if (n < 2) return 0.0;
    std::vector<double> active(n);
    for (std::size_t i = 0; i < n; ++i) active[i] = r[i] - b[i];
    double mean = std::accumulate(active.begin(), active.end(), 0.0) / n;
    double var  = 0.0;
    for (auto x : active) { double d = x - mean; var += d * d; }
    double te = std::sqrt(var / (n - 1)) * std::sqrt(TRADING_DAYS);
    return te > 0 ? mean * TRADING_DAYS / te : 0.0;
}

double PerformanceCalculator::tracking_error(const std::vector<double>& r,
                                              const std::vector<double>& b) {
    std::size_t n = std::min(r.size(), b.size());
    if (n < 2) return 0.0;
    std::vector<double> active(n);
    for (std::size_t i = 0; i < n; ++i) active[i] = r[i] - b[i];
    double mean = std::accumulate(active.begin(), active.end(), 0.0) / n;
    double var  = 0.0;
    for (auto x : active) { double d = x - mean; var += d * d; }
    return std::sqrt(var / (n - 1)) * std::sqrt(TRADING_DAYS);
}

// ─── Print ────────────────────────────────────────────────────────────────────
void PerformanceCalculator::print(const Metrics& m, const std::string& title) {
    spdlog::info("========== {} ==========", title);
    spdlog::info("Total Return      : {:.2f}%", m.total_return        * 100);
    spdlog::info("CAGR              : {:.2f}%", m.cagr                * 100);
    spdlog::info("Annualised Vol    : {:.2f}%", m.annualised_vol      * 100);
    spdlog::info("Max Drawdown      : {:.2f}%", m.max_drawdown        * 100);
    spdlog::info("Sharpe Ratio      : {:.3f}",  m.sharpe);
    spdlog::info("Sortino Ratio     : {:.3f}",  m.sortino);
    spdlog::info("Calmar Ratio      : {:.3f}",  m.calmar);
    spdlog::info("Win Rate          : {:.2f}%", m.win_rate            * 100);
    spdlog::info("Profit Factor     : {:.3f}",  m.profit_factor);
    spdlog::info("VaR 95%           : {:.2f}%", m.var_95              * 100);
    spdlog::info("CVaR 95%          : {:.2f}%", m.cvar_95             * 100);
    spdlog::info("Alpha             : {:.2f}%", m.alpha               * 100);
    spdlog::info("Beta              : {:.3f}",  m.beta);
    spdlog::info("Information Ratio : {:.3f}",  m.information_ratio);
    spdlog::info("Tracking Error    : {:.2f}%", m.tracking_error      * 100);
    spdlog::info("==================================");
}

// ─── JSON serialisation ───────────────────────────────────────────────────────
std::string PerformanceCalculator::to_json(const Metrics& m) {
    nlohmann::json j = {
        {"total_return",       m.total_return},
        {"annualised_return",  m.annualised_return},
        {"cagr",               m.cagr},
        {"annualised_vol",     m.annualised_vol},
        {"max_drawdown",       m.max_drawdown},
        {"sharpe",             m.sharpe},
        {"sortino",            m.sortino},
        {"calmar",             m.calmar},
        {"win_rate",           m.win_rate},
        {"profit_factor",      m.profit_factor},
        {"avg_win",            m.avg_win},
        {"avg_loss",           m.avg_loss},
        {"var_95",             m.var_95},
        {"cvar_95",            m.cvar_95},
        {"skewness",           m.skewness},
        {"kurtosis",           m.kurtosis},
        {"alpha",              m.alpha},
        {"beta",               m.beta},
        {"information_ratio",  m.information_ratio},
        {"tracking_error",     m.tracking_error},
        {"n_trades",           m.n_trades},
        {"n_bars",             m.n_bars},
    };
    return j.dump(2);
}

}  // namespace estateiq
